### ISSUE TITLE
False positive "for loop analysis" warning

### ISSUE THREAD

**Issue by rHermes** [llvm/llvm-project#109402]
This is tested with clang 18.1.0 on godbolt and occurs on trunk.

Here is the godbolt: https://godbolt.org/z/o3jhnoPso

Here is the code:
```c++
#include <iostream>
#include <vector>

/**
 * Definition for singly-linked list.
 */
struct ListNode
{
  int val;
  ListNode* next;
  ListNode() : val(0), next(nullptr) {}
  ListNode(int x) : val(x), next(nullptr) {}
  ListNode(int x, ListNode* next) : val(x), next(next) {}
};

class Solution
{
public:
  static std::vector<std::vector<int>> spiralMatrix(const int M, const int N, ListNode const* head)
  {
    std::vector<std::vector<int>> out(M, std::vector<int>(N, -1));

    const auto writeVal = [&](const int y, const int x) {
      if (head) {
        out[y][x] = head->val;
        head = head->next;
      }
    };

    // ok, we will loop in numbers.
    for (int iter = 0; head; iter++) {
      const int W = N - 2 * iter;
      const int H = M - 2 * iter;

      for (int j = 0; j < W; j++) {
        writeVal(iter, iter + j);
      }

      for (int j = 1; j < H; j++) {
        writeVal(iter + j, iter + W - 1);
      }

      for (int j = W - 2; 0 <= j; j--) {
        writeVal(iter + H - 1, iter + j);
      }

      for (int j = H - 2; 0 < j; j--) {
        writeVal(iter + j, iter);
      }
    }

    return out;
  }
};
```

If you compile this like so you get the following warning:
```c++
clang++ -Wall -Wextra -std=c++2a -Werror -o test test.cpp
<source>:31:24: error: variable 'head' used in loop condition not modified in loop body [-Werror,-Wfor-loop-analysis]
   31 |     for (int iter = 0; head; iter++) {
      |                        ^~~~
```

**Comment by i42output** [comment]
Simpler testcase exhibiting the same behaviour:

```
int main()
{
    bool xyzzy = true;

    auto foo = [&]()
    {
        thread_local int i = 0;
        xyzzy = (++i == 10);
    };

    for (int i = 0; xyzzy; ++i)
    {
        foo();
    }
}
```

**Comment by EugeneZelenko** [comment]
Could you please try 19 or `main` branch? https://godbolt.org should be helpful.

**Comment by rHermes** [comment]
I am getting the same error on trunk and 19 :)

**Comment by shafik** [comment]
looks like this goes back all the way to clang-3.8: https://godbolt.org/z/66oxxG9oW

**Comment by RedBeard0531** [comment]
We just hit the same issue with clang-19, and it seems to repro with 21 and truck (as of today): https://godbolt.org/z/qh6GTK8Yb

**Comment by RedBeard0531** [comment]
Note that it should also treat any black-box call as a usage if the variable has escaped, to avoid false positives, at least in the version enabled by `-Wall`. I could imagine some people wanting a stricter version that warns on the "I can't prove whether it was modified or not" case, but that seems a bit aggressive for something in `-Wall`
